ODIN_II/SRC/parse_making_ast.cpp

```diff
@@ -1355,7 +1362,14 @@ ast_node_t* newMultipleInputsGateInstance(char* gate_instance_name, ast_node_t*
         symbol_node = newSymbolNode(gate_instance_name, line_number);
     }
 
-    char* newChar = vtr::strdup(expression1->types.identifier);
+    char* newChar;
+    if (expression1->types.identifier) {
+        newChar = vtr::strdup(expression1->types.identifier);
+    } else {
+        oassert(expression1->num_children > 0);
+        oassert(expression1->children[0]->type == IDENTIFIERS && "Expected child to be an identifier");
```

Comment:
I think it would be better to be a separate function -> get_instance_port_identifier()
in it I would check if it is a RANGE_REF or an ARRAY_REF then look for the symbol at ->children[0]

the assert are not essential since it should not get there unless the bison is messed up.

My biggest worry is
```
`define REVERSE 0
module and2 (in1,in2,out);
input  [1:0] in1,in2;
output [1:0] out;

and((REVERSE)?out[0]:out[1], in1[0],in2[0]);
and(out[1], in1[1],in2[1]);
endmodule 
```

this should be allowed if i am not mistaken ![Verilog 2005](http://staff.ustc.edu.cn/~songch/download/IEEE.1364-2005.pdf) but correct me if i am wrong

Reply:
I think it is better to leave the tree as is at that stage, and only check the port name when it is needed, we can build the circuitry later if we can't resolve the condition (a mux) or it will be folded in one of the phases.

Reply:
My opinion is that asserts are essential whenever there is a sentence that starts with 'we should not get here unless...' (that's kinda the point of them)

Reply:
> I think it is better to leave the tree as is at that stage, and only check the port name when it is needed, we can build the circuitry later if we can't resolve the condition (a mux) or it will be folded in one of the phases.

Do you mean that we should get rid of the check

```
for (long j = 0; j < list_of_module_items->num_children && variable_found == false; j++) {
    if (list_of_module_items->children[j]->type == VAR_DECLARE_LIST) {
        for (long k = 0; k < list_of_module_items->children[j]->num_children; k++) {
            if (strcmp(list_of_module_items->children[j]->children[k]->children[0]->types.identifier, module_variables_not_defined[i]->children[0]->children[0]->types.identifier) == 0)
                variable_found = true;
        }
    }
}
if (!variable_found)
    add_child_to_node_at_index(list_of_module_items, module_variables_not_defined[i], 0);
else
    free_whole_tree(module_variables_not_defined[i]);
```
and then do that later?

Or do you mean merge this fix and then make it work for more complicated expressions later?

Reply:
correct.
[IEEE.1364-2005](http://staff.ustc.edu.cn/~songch/download/IEEE.1364-2005.pdf):: pg 173 
section **12.3 Ports** it shows concat and say any valid expression as port.

expecting a certain subtree and doing constant folding on ternary etc.. becomes a bit more involved for this phase (parsing into an ast).
ast optimization should fold any foldable logic and leave you with something you can work with after.
At that point, if the subtree is not just a symbol, array_ref or range_ref (concatenations may resolve weirdly... I remember this being an issue) you can build logic and do hookup directly since the names would be for the subexpressions rather than the port directly.

Reply:
I can work with this as a bug fix, but a proper fix is still needed (and should be used by most of everything using ports).
this will also simplify things when we work on scoping issues

Reply:
Ok so should I raise a separate issue for compound statements as the output for a port. And we can merge this to fix for vector types?

Reply:
correct.
It does fix the initial issue and it is an improvement.

as for the regression diffs,  #1247 should fix parts of the problem